File: spark16/src/main/java/javaudf/StringGroupCount.java
package javaudf;

import org.apache.spark.sql.Row;
import org.apache.spark.sql.expressions.MutableAggregationBuffer;
import org.apache.spark.sql.expressions.UserDefinedAggregateFunction;
import org.apache.spark.sql.types.DataType;
import org.apache.spark.sql.types.DataTypes;
import org.apache.spark.sql.types.StructField;
import org.apache.spark.sql.types.StructType;

import java.util.ArrayList;
import java.util.List;

import static org.apache.spark.sql.types.DataTypes.IntegerType;

/**
 * Created by shengjk1 on 2016/8/9.
 */
public class StringGroupCount extends UserDefinedAggregateFunction {
	@Override
	public StructType inputSchema() {
		List<StructField> fle=new ArrayList<>();
		fle.add(DataTypes.createStructField("name",DataTypes.StringType,true));
		
		return DataTypes.createStructType(fle);
	}
	
	@Override
	public StructType bufferSchema() {
		List<StructField> fle=new ArrayList<>();
		fle.add(DataTypes.createStructField("count", IntegerType,true));
		
		return DataTypes.createStructType(fle);
	}
	
	@Override
	public DataType dataType() {
		return IntegerType;
	}
	
	@Override
	public boolean deterministic() {
		return true;
	}
	
	@Override
	public void initialize(MutableAggregationBuffer buffer) {
		buffer.update(0,0);
	}
	
	//中间计算结果
	@Override
	public void update(MutableAggregationBuffer buffer, Row input) {
		buffer.update(0,buffer.getInt(0)+1);
	}
	
	//对中间结果合并
	@Override
	public void merge(MutableAggregationBuffer buffer1, Row buffer2) {
		buffer1.update(0,buffer1.getInt(0)+buffer2.getInt(0));
	}
	
	//最终返回结果
	@Override
	public Object evaluate(Row buffer) {
		return buffer.get(0);
	}
}

File: spark16/src/main/java/sudf/StringGroupCount.scala
package sudf

import org.apache.spark.sql.Row
import org.apache.spark.sql.expressions.{MutableAggregationBuffer, UserDefinedAggregateFunction}
import org.apache.spark.sql.types._

/**
  * Created by root on 2016/8/8.
  */
class StringGroupCount extends UserDefinedAggregateFunction{
  override def inputSchema: StructType =
    StructType(Array(StructField("str",StringType,true)))

  //中间结果
  override def bufferSchema: StructType = {
    StructType(Array(StructField("count",IntegerType,true)))
  }

  override def dataType: DataType = IntegerType

  override def deterministic: Boolean = true

  override def initialize(buffer: MutableAggregationBuffer): Unit =
    buffer(0)=0

  override def update(buffer: MutableAggregationBuffer, input: Row): Unit = {
    buffer(0)=buffer.getAs[Int](0)+1
  }

  override def merge(buffer1: MutableAggregationBuffer, buffer2: Row): Unit = {
    buffer1(0)=buffer1.getAs[Int](0)+buffer2.getAs[Int](0)
  }

  override def evaluate(buffer: Row): Any = {
    buffer.getAs[Int](0)
  }
}
